\documentclass[a4paper,12pt]{report}

%packages
\usepackage[utf8]{inputenc}
\usepackage{mathtools}
\usepackage[onehalfspacing]{setspace}

%document
\begin{document}
\begin{titlepage} 
  \parindent0pt 
  \hskip.25cm \vrule width.8pt \hskip.25cm 
  \begin{minipage}{\dimexpr\linewidth*2-2.5cm-.8pt-2.5cm\relax} 
    \sffamily\bfseries {
    \huge Master 2 Bioinformatique\par}\large 
    Semestre 10\par Université de Bordeaux\par 
    %%Année truc
  \end{minipage}%
  \vskip10pt plus 1fil
  \fbox{%
  \begin{minipage}{\dimexpr\linewidth-2\fboxsep-2\fboxrule} 
    \centering\Large\bfseries 
    \vskip1cm 
    Project report: TITLE\par 
    \vskip1cm \kern0pt 
  \end{minipage}%
  }%
  \vskip0.5cm 
  {\centering 
  \vskip0pt plus 1fil 
  }%

  \textbf{Aurélien \sc{Luciani}}\centering\\ 
  \vskip0pt plus 1fil 
  \hfill \hskip.25cm \vrule width.8pt \hskip.25cm 
  \vskip0pt plus 1fil {\centering Supervisor\,: Murray P. \sc{Cox} \par }
  \vskip0pt plus 2fil 
  \hfill\bfseries{Year 2014-2015}\hfill\null 
\end{titlepage}

% to be removed...
%\setcounter{tocdepth}{1}
% ...until here
\tableofcontents

\chapter*{Introduction}
\addcontentsline{toc}{chapter}{Introduction}
Two groups of populations can be identified in the Islands of South-East Asia (ISEA), one is composed of the Melanesians, whose ancestors settled in these islands during the first human settlement, around 45 thousand years ago. The other arrived more recently during a period often called the Austronesian expansion, between 5 and 4 thousand years ago, when people from mainland China settled in the islands. Nowadays people living in this area have mixed genomic ancestry and markers can be identified and defined as either from an Asian ancestry or a Melanesian one. These markers are based on signle nucleotid polymorphisms located in different chromosomes and 52 markers can be used to define accurately the admixture of the Asian ancestry in every individuals \cite{Cox01}.

The choice of these SNPs is a result of previous studies at Massey University, the University of Arizona, the Santa Fe Institute, and the Eijkman Institute that sequenced 1430 individuals from 60 populations. This set of SNPs is defined as highly informative and allows to define the ancestry of a person based on a small quantity of markers that are highly discriminant.

Two specific patterns can be seen, one is the non linear gradient of Asian admixture when observing individuals in the different islands when looking along the longitudinal axis, corresponding more or less the wave the settling might have happened. The second is the difference of admixture when looking at specific parts of the genomes associated with male or female ancestry, implying a gender-biased expansion.
% first papers
% whole project for the group
%% leading to the creation of the model


\chapter{Project presentation}
The project consists of developing a model of the Austronesian expansion throughout the ISEA that could reproduce the same two patterns observed in the real data.
The first pattern can be seen in the figure (REF!!!) and is a non-linear gradient in Asian admixture that declines abruptly around the eastern part of Indonesia. The second one, represented in the figure (REF!!!), is a 

\section{Type of values observed}

\section{Model used}
\subsubsection{Comparison to other models}
\subsubsection{Parameters}
% model details
% previous works
%% previous models
%% previous analyses

\section{Statistical analysis framework}
\subsubsection{Previous works}
\subsubsection{Specific needs}


\chapter{Implementation}
% or "realisation", "analysis pipeline"

% run management (local, cluster, NeSI)
\section{Run management}
The fact that many simulations are required to be able to infer meaningful information from the model imply that they cannot be run on a single desktop computer. In fact, running them in a single computer, the time to run enough simulations to be statistically relevant could be counted in years.
Luckily, every simulation being independent from the others, there are multiple ways to generate more results in less time.
Firstly, since nowadays most of the computers have multiple processors, one computer can run multiple simulation concurrently. Also, a cluster of computers can be used and the wanted simulations can be dispatched among the nodes of the cluster so that they each run the simulations they were assigned and when all the nodes have ended their runs, their outputs can be aggregated and/or stored

Since running a lot of simulations, with a lot of different parameters, can generate a lot of output files whose original parameters can be hard to track, a way to keep them organised was needed. Doing so, it would also be possible to use results from different batches and to analyse them together, thus avoiding to redo simulations for parameters already tested.

The organised way to store this is naturally in a database. The choice has been made to use a relational database, specifically MariaDB (a fork of Oracle's MySQL). The structure has been designed so that the parameters can be efficiently queried. The tables and relations between them can be seen in the figure \ref{db}. It has first been developed and tested locally and then, once it was working properly, it was deployed on a server provided by Massey University for research purposes.

\begin{figure}
	\centering
	\includegraphics[scale=0.7]{data/DB.png}
	\caption{Structure of the database storing the results of the simulations and the corresponding parameters}
	\label{db}
\end{figure}

% data processing from model output
% data storage, adding and querying data
\section{Data processing, storage and query}


\section{Analysis}


\subsection{Comparisons}
To be able to compare two different scenarios, one has to define comparison functions that will be able to provide a value of similarity or dissimilarity when provided with observable values that characterise the scenarios. In this case, the comparison is done between the real observed data and the output of one simulation.
Every simulation output has to be treated to be of the same form than the observed data so that it can be compared. The observed data does not include values for specific islands that are included in the simulations, namely Alor, Tanimbar and Aru in eastern Indonesia. The admixture values for mitochondrion and Y-chromosome are also not available for every island.
Those values in the models can then be dropped as they cannot be compared.

Many different comparison functions have been tested. Two functions have been selected that can hold different information about a comparison. The mean square distance (MSD) and the partial Mantel test have been selected as they do not have a good correlation when applying them to random admixture values, meaning that they do not carry the same information, they are complementary, and using them both gives a better overview of the comparison.

\subsubsection{Mean square distance}
The mean square distance is the mean value of all the distances between values of admixture for every islands.

\begin{equation}
MSS = \frac{\sum\limits_{i=1}^{n} (Real_i - Simulated_i)) ^ 2}{n}
\end{equation}

It gives a distance value, with 0 meaning that the two observed values are absolutely identical. In this specific context, with the fact that an admixture value is compared and because of the values observed in the real data, the higher possible value will be around $0.8$.


\subsubsection{Partial Mantel test}
The Mantel test has been developed to be able to compare two matrices with the same information, the partial version of this test also uses a third matrix, holding geographical distances for the cells of the matrices to be able to weight the values according to the actual geographical distance of the points \cite{Smo01}.

In this case, the matrices contain the values of distances of admixtures between every islands in the graph and a matrix $M$ is calculated as
\begin{equation}
	M = \begin{bmatrix}
		d(A_0, A_0)	& \cdots & d(A_0, A_n) 	\\
		\vdots		& \ddots & \vdots		\\
		d(A_n, A_0) & \cdots & d(A_n, A_n)
	\end{bmatrix}
\end{equation}
were $d$ is the function returning the distance between the two arguments and $A$ the admixtures of the $n$ islands of the graph.

the partial Mantel test returns a correlation value that is between -1 and 1 with a value of 0 meaning that the two matrices are not correlated at all and 1 that they are completely correlated.

\subsection{Visualisations}

% ABC framework



\chapter{Results}
\section{Grid search analysis}

\subsubsection{Stability analysis}


\subsubsection{Sensitivity analysis}


\section{ABC Framework}

\subsection{Priors definition}


\subsection{Posterior distribution}




\chapter{Discussion}
% optimisations
\section{Optimisations}
Some key steps in the analysis framework need to be efficient enough so that time is not lost waiting for results to be treated or for graphs to be plotted and that the computation can be done without needing a computer.
This project has seen a few important refactoring processes to be able to cope with the quantity of data to be treated. The most important have been to treat the simulation results as a stream of values instead of loading the whole dataset in memory. This changed the memory complexity from linear to constant and it actually improved the time complexity from sub-quadratic to linear for the treatment part and from sub-quadratic to linearithmic for the analysis part. The improvement in time has been made possible by assuming that the results of one simulation are always together in the result stream, that way saving the cost of searching results in a big block of memory when they are actually next to each other.
Actual execution times have been recorded and can be seen in the graph (REF!!!).\

The stream approach also allows the different steps of an analysis to be run simultaneously, by piping each step to the next, effectively making the whole process run in parallel. This is useful only if the computer used has at least $n$ cores if $n$ processes need to be run in parallel. In this case, the time of the whole process is the time needed to run the longest step.


% other visualisations
\section{Other visualizations}
% more complex model
\section{More complex model}


\chapter*{Conclusion}
\addcontentsline{toc}{chapter}{Conclusion}

\bibliography{main}
\bibliographystyle{apalike}

\appendix
\addtocontents{toc}{\protect\setcounter{tocdepth}{-1}}
% example visualisations
%\addtocontents{toc}{\setcounter{tocdepth}{-1}}
\chapter{Examples of visualisation}
%\addtocontents{toc}{\protect\setcounter{tocdepth}{0}}

\section{Stability}
\section{Sensitivity}

% overview

\end{document}


